ublox_derive/src/output.rs

```diff
@@ -0,0 +1,140 @@
+use crate::types::{PackDesc, PackFieldEnum, PackFieldFlags, PackFieldRepr};
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+use syn::Type;
+
+pub fn generate_recv_code_for_packet(pack_descr: &PackDesc) -> TokenStream {
+    let ref_name = format_ident!("{}Ref", pack_descr.name);
+    let mut getters = Vec::with_capacity(pack_descr.fields.len());
+
+    for f in &pack_descr.fields {
+        let ty = f.intermidiate_type();
+        let get_name = &f.name;
+        getters.push(quote! {
+            #[inline]
+            pub fn #get_name(&self) -> #ty {
+            unimplemented!()
+            }
+        });
+    }
+    quote! {
+    pub struct #ref_name<'a>(&'a [u8]);
+    impl<'a> #ref_name<'a> {
+            #(#getters)*
+    }
+    }
+}
+
+pub fn generate_types_for_packet(pack_descr: &PackDesc) -> TokenStream {
+    let mut req_types = Vec::new();
+    for f in &pack_descr.fields {
+        let ty = f.intermidiate_type();
+        match f.repr {
+            PackFieldRepr::Plain | PackFieldRepr::Map(_) => {}
+            PackFieldRepr::Enum(ref e) => {
+                req_types.push(define_enum(e, ty, &f.ty));
+            }
+            PackFieldRepr::Flags(ref flags) => {
+                req_types.push(define_bitflags(flags, ty, &f.ty));
+            }
+        }
+    }
+    let name = format_ident!("{}Raw", pack_descr.name);
+    let class = pack_descr.header.class;
+    let id = pack_descr.header.id;
+    let fixed_len = match pack_descr.header.fixed_len {
+        Some(x) => quote! { Some(#x) },
+        None => quote! { None },
+    };
+    quote! {
+        #(#req_types)*
+
+    impl UbxPacket for #name {
+        const class: u8 = #class;
+        const id: u8 = #id;
+    const fixed_length: Option<u16> = #fixed_len;
+    }
+    }
+}
+
+pub fn generate_send_code_for_packet(pack_descr: &PackDesc) -> TokenStream {
+    let payload_struct = format_ident!("{}Payload", pack_descr.name);
+
+    let mut fields = Vec::with_capacity(pack_descr.fields.len());
+    for f in &pack_descr.fields {
+        let ty = f.intermidiate_type();
+        let name = &f.name;
+        fields.push(quote! {
+            #name: #ty
+        });
+    }
+
+    match pack_descr.packet_size() {
+        Some(packet_size) => {
+            quote! {
+            pub struct #payload_struct {
+                #(#fields),*
+            }
+            impl #payload_struct {
+                #[inline]
+                fn to_packet(self) -> [u8; #packet_size] {
```

Comment:
For the fixed-size case, should we instead of returning a value, take a `&mut [u8; #packet_size]`, so that a user can populate a packet into a buffer that they already have?

Continuing on this, a user may not want to instantiate an extra struct, if they already have a transmit buffer.

Reply:
>  so that a user can populate a packet into a buffer that they already have?

This would be not nice interface I suppose.
User's buffer can be not exactly the same size. In this case it would be impossible to get `&mut [u8; #packet_size]` without unsafe, `&mut [u8]` can not be converted to `&mut [u8; #packet_size]` without unsafe.

Also from compiler with optimization point of view this doesn't help a lot.
```rust
#[inline]
fn foo(s: S) -> [u8; X] { ... }
let mut my_buf[u8; X];
my_buf = foo(S {});
```
After inline `foo` `s: S` should be  copied exactly into `my_buf` without 
usage of intermediate buffer, something like  "copy elision elimination" from C++.

Actually the idea for sender that compiler eliminate almost all stuff and it would be like this:
```
struct PackPayoud {...}
impl PackPayoud {
   pub fn to_packet(self) -> [u8; X] { ... }
}

let buf = PackPayload { x1: f1, x2: f2}.to_packet();
```
->
```rust
let mut buf: [u8; X];
buf[..] = init_header;
buf[6] = to_bits(f1);
buf[end] = calc_crc();
```
So at the end it would be code without any copy data around.

 > Continuing on this, a user may not want to instantiate an extra struct, if they already have a transmit buffer.

I don't consider this use-case, because of I think user can only send "configure" packages.
And I don't see much sense for this use-case to change configuration many times times like this:
```rust
cfg1.pref1 = 1;
cfg1.pref1 = 2;
cfg1.pref1 = 3;
```
I suppose configuration will be settled only once after start of program and that's all.

For this use-case I suppose there is need for your `Setter`, but what use-case for this,
what packet user have to send in this way?

Reply:
> In this case it would be impossible to get &mut [u8; #packet_size] without unsafe

I believe try_into works for this:
```
use std::convert::TryInto;

pub fn bar(data: &mut [u8; 12]) {
    data[0] = 5;
}

pub fn foo(data: &mut [u8]) {
    bar(data.try_into().unwrap());
}
```
Of course, I unwrap there, but you should return the Result on error.

> For this use-case I suppose there is need for your Setter, but what use-case for this,
what packet user have to send in this way?

I had imagined a scenario where e.g. you have a DMA buffer for your UART, which you've already allocated for previous transfers. For example, some chips in [NXP's LPC](https://community.nxp.com/docs/DOC-334851) series allow DMA to the UART (not any that I'm using, but I would if I could).

Even if you haven't already allocated the DMA buffer, you still need to pin it in memory somewhere, so that you can pop the stack before the DMA completes.

I agree that the usecase of "I have a particular config packet which I need to send multiple times with slight variations" is unusual.

Reply:
It may be that with the compiler optimizations you can still call `to_packet` into a local variable, then copy the local variable into the DMA buffer and have it still be zero copy. However, the code doesn't "look" zerocopy, so it takes more brainpower for me to think "oh, this code is actually pretty fast."

Reply:
I suppose in this case it is better to accept `std::io::Write` as argument.
It is already implemented for `Vec<u8>` and it is more nice way to do things,
for example user can reallocate and pin DMA buffer if there are no space.
And common users can just pass `Vec<u8>` and care about ask about suitable size,
allocate size and that call `foo` ?

Reply:
Secretly, I want to permit `no_std` in the future, so `std::io::Write` might be a bit heavyweight. Also it might not still be zerocopy? I don't know.

Reply:
> Secretly, I want to permit no_std in the future, so std::io::Write might be a bit heavyweight. Also > it might not still be zerocopy? I don't know.

We actually need one function, so I suppose it is possible to define it in such way for no_std case:
```
#[error]
enum NotEnoughMem {}

trait MyWrite {
    fn write(&mut self, buf: &[u8]) -> Result<usize, NotEnoughMem>;
}

impl MyWrite for Vec<u8> {
    fn write(&mut self, buf: &[u8]) -> Result<usize, NotEnoughMem> {
        unimplemented!();
    }
}

impl<'a> MyWrite for &'a mut [u8] {
    fn write(&mut self, buf: &[u8]) -> Result<usize, NotEnoughMem> {
        unimplemented!();
    }
}
```